plugins/reporters/spdx/src/main/kotlin/SpdxDocumentModelMapper.kt

```diff
@@ -64,21 +65,21 @@ internal object SpdxDocumentModelMapper : Logging {
 
         val projects = ortResult.getProjects(omitExcluded = true, includeSubProjects = false).sortedBy { it.id }
         val projectPackages = projects.map { project ->
-            val spdxProjectPackage = project.toPackage().toSpdxPackage(
+            project.toPackage().toSpdxPackage(
                 SpdxPackageType.PROJECT,
                 licenseInfoResolver,
                 ortResult
             )
+        }
 
-            ortResult.getDependencies(project.id, 1).mapTo(relationships) { dependency ->
```

Comment:
Just wondering, can't the same bug also appear around line 94, if a non-excluded package depends on an excluded package?

In general, I wonder whether a more general solution to the problem would be to also teach `getDependencies()` an `omitExcluded` parameter. If we had that, the fix would be a matter of just setting `omitExcluded = true` in current line 73 and 94, right?

Reply:
```
root-project-1
|
+- sub-project-1
   |
   +- package (dependency of "sub-project" only)
root-project-2
|
+- sub-project-2
   |
   +- package (dependency of "sub-project" only)
```

If we have above, and consider the original (current) algorithm implemented by you, I guess the intention was to not mention subproject-1 and subproject-2 in the SPDX report, but link the dependencies directly to root-project-1 and root-project-2.

To my understanding the implementation did not match this mentioned intention, which I wanted to fix with this new implementation.

> are you saying there should be no relationship between "sub-project" and "package" being recorded in the SPDX document?

Given the above, I'd say no, because "package" should be a direct child of the "root" project, and nested submodules should not be modeled in the output.

Reply:
> I guess the intention was to not mention subproject-1 and subproject-2 in the SPDX report, but link the dependencies directly to root-project-1 and root-project-2.

Actually no, that was no my intention. The intention was to create all of the following relationships in this case:

- `root-project-1` depends on `sub-project-1`
- `sub-project-1` depends on `package 1`

(And the same for the number 2 projects / packages.)

Reply:
This surprises me, because:

1. https://github.com/oss-review-toolkit/ort/commit/b47154482299b9b49c0e0c692e3f2dacdd53c395
    The commit message clearly says "...but to the real root projects"
2. That is reflected by the limitation of the depth to 1 here: https://github.com/oss-review-toolkit/ort/commit/b47154482299b9b49c0e0c692e3f2dacdd53c395#diff-a7d33dd6b251d36d6f79411935a6f2e35586240fa8e3337cb29b1c7a9b89403dR76

Reply:
But no matter what the original intention of that commit was, I believe we should do now what we think is the best.
In my opinion, only listing the top-level projects is better than exposing the whole sub-module structure. Currently, only top-level projects are exposed, so I would prefer to not start exposing sub-module structure of all nesting depths, but to stick with top level project and fix the underlying tree accordingly. 

Would that be ok with you?

Reply:
> The commit message clearly says "...but to the real root projects"

The "but to the real root projects" part referred to the goal to link direct dependencies to the project they belong to, and not "to an arbitrary root project" like it was before.

> That is reflected by the limitation of the depth to 1 here

The idea behind that was the following: It is enough for each project to only link to its direct dependencies (with depth == 1), as in the other loop below, each package in turn gets linked to *its* direct dependencies. So even if each entity (project or package) only relates to its direct dependencies, in the end we get a full dependency tree in total.

Reply:
> Well, that differs from what @tsteenbe and others have expressed [here](https://github.com/oss-review-toolkit/ort/issues/4099#issuecomment-866253887) then. The intent of my [b471544](https://github.com/oss-review-toolkit/ort/commit/b47154482299b9b49c0e0c692e3f2dacdd53c395) was to partly address that linked issue.

I understand Thomas' comment differently, however. It's not mentioning sub-modules explicitly, so this is IMO speculation. I believe we should discuss sub-module case explicitly / separately.

> And we have the https://github.com/oss-review-toolkit/ort/issues/3906, BTW.

Ok.

> Not really, to be honest, as that takes us a step back again

Why is it a step back. My PR does not reduce information reported at all. But fixes this bug:

> However, I completely agree that I've introduced a bug in https://github.com/oss-review-toolkit/ort/commit/b47154482299b9b49c0e0c692e3f2dacdd53c395 by disregarding excludes.

Reply:
> So even if each entity (project or package) only relates to its direct dependencies, in the end we get a full dependency tree in total.

I'm not sure if this bug was clear to you already, but this sentence illustrates it pretty well. The idea does not work in case there are sub-projects, because then one does not iterate over the sub-projects dependencies at all in some cases.

Reply:
> ```
> root-project-1
> |
> +- sub-project-1
>    |
>    +- package 1
> root-project-2
> |
> +- sub-project-2
>    |
>    +- package 2  
> ```
> 
> [...]
>
> Given the above, I'd say no, because "package" should be a direct child of the "root" project, and nested submodules should not be modeled in the output.

To me it would be misleading to report "package-1" as a direct dependency of "root-project-1" here, especially if both "root-project-1" and "sub-project-1" are published. I do not yet understand the issue with the current approach to build the tree, especially:

> The idea does not work in case there are sub-projects, because then one does not iterate over the sub-projects dependencies at all in some cases.

Under which circumstances would that be the case?

Reply:
@sschuberth  @mnonnenmacher  - 

I've adjusted the implementation to what I understood We agreed upon in todays call.
@sschuberth  I've incorporated your proposal from the PR comments to teach `getDependencies()` to omit excluded IDs.